File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 1)return nums[0];
        vector<int> dp(nums.size() , 0);
        int n = nums.size();
        dp[0] = nums[0];
        dp[1] = nums[1];
        for(int i = 2 ; i < n ; i++){
            int left = (i-3 >= 0)?dp[i-3]:INT_MIN;
            int right = dp[i-2];
            dp[i] = nums[i] + max(left , right);
        }

        return max(dp[n-1] , dp[n-2]);
    }
};

File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        dp = [-1]*len(nums)
        for i in range(len(nums)):
            prev1 = dp[i-3] if i-3 >=0 else 0
            prev2 = 0 if i - 2 < 0 else dp[i-2]
            dp[i] = max(prev1,prev2) + nums[i]
        print(dp)
        return max(dp[-1],dp[-2])